Change AssemblyName formatting not to require static table of KVPs

marek-safar opened the pull request:
to do loop with span operations to replace single characters

Expanded test coverage to actually cover this odd behaviour

marek-safar committed:
Change AssemblyName formatting not to require static table of KVPs
to do loop with span operations to replace single characters

Expanded test coverage to actually cover this odd behaviour

danmosemsft commented on src/libraries/System.Reflection/tests/AssemblyNameTests.cs:
```diff
@@ -34,6 +34,22 @@ public static IEnumerable<object[]> Names_TestData()
             yield return new object[] { "\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B", "\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B" };
         }
 
+        public static IEnumerable<object[]> Names_TestDataQuoted()
```
Why quoted -- would something like `Names_TestDataRequiresEscaping()` be better?

marek-safar replied:
Updated the name

danmosemsft commented on src/libraries/System.Reflection/tests/AssemblyNameTests.cs:
```diff
@@ -34,6 +34,22 @@ public static IEnumerable<object[]> Names_TestData()
 
+        public static IEnumerable<object[]> Names_TestDataQuoted()
+        {
+            yield return new object[] { " name ", "\" name \"" };
```
Did you miss the case of escaping a backslash?

marek-safar replied:
Included

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
+                        sb.Append('\\');
+                        break;
+                    case '\t':
+                        sb.Append('\\').Append('t');
```
Is this faster than `sb.Append("\\t")`?

marek-safar replied:
It's probably same now when Append is has a fast path for 2 characters string (updating)

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
                         continue;
-                    if ((s.Length - i) < escapeReplacement.Length)
+                    case '\n':
+                        sb.Append('\\').Append('n');
```
Same question.  Wondering why we use the longer variant.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
+
+                        --i;
+                        break;
+#endif
```
Is this a change in behavior?  Seems like a \n on Windows (no proceeding \r) would have still resulted in a \r\n, but now it's only going to be a \n?

stephentoub replied:
And previously \r wasn't special-cased if it was on its own, was it?

marek-safar replied:
Good point simplified it back

marek-safar committed:
Review feedback

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:
```diff
@@ -99,24 +99,36 @@ private static void AppendQuoted(this StringBuilder sb, string s)
+                        sb.Append("\\t");
                         continue;
-                    if ((s.Length - i) < escapeReplacement.Length)
+#if TARGET_WINDOWS
```
This Windows specific behavior sounds like a bug introduced in the managed rewrite of AssemblyName formatters. It does not really make sense for the AssemblyName escaping rules to differ between Windows and Unix.

I would recommend to change the managed formatter to do what the unmanaged formatter has been always doing: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/binder/textualidentityparser.cpp#L740

marek-safar replied:
Done.

marek-safar committed:
Realign the implementation with original code at

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/binder/textualidentityparser.cpp#L690
